### PR TITLE
TMDM-14192 AUTO_INCREMENT value for NOT Primary Key reset to 0 after MDM restart

### PR DESCRIPTION
**Pull request by zfkang-talend**
JIRA: https://jira.talendforge.org/browse/TMDM-14192

**What is the current behavior?** (You should also link to an open issue here)
1. For SOAP/REST API, one AUTO_INCREMENT field's value cann't saved after restart MDM server.
2. For Bulkload, don't support the AUTO_INCREMENT field(Not PK).

**What is the new behavior?**
1. Persisted to the database for AUTO_INCREMENT field's value.
2. Support AUTO_INCREMENT field in Bulkload

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core/src/com/amalto/core/load/LoadParser.java [hunk 1]
```diff
@@ -146,18 +150,34 @@ public static StateContext parse(InputStream inputStream, Configuration config,
     public static class Configuration {
         private final String payLoadElementName;
         private final String[] idPaths;
+        private final String[] normalFieldPaths;
         private final boolean autoGenPK;
         private final String dataClusterName;
         private final String dataModelName;
         private final AutoIdGenerator idGenerator;
+        private final AutoIdGenerator[] normalAutoFieldGenerator;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/LoadParser.java]: Change to `normalFieldGenerator`

File: org.talend.mdm.core/src/com/amalto/core/load/action/OptimizedLoadAction.java [hunk 2]
```diff
@@ -37,54 +37,91 @@
     private final String typeName;
     private final String dataModelName;
     private final boolean needAutoGenPK;
+    private final boolean needAutoGenAutoFields;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/action/OptimizedLoadAction.java]: Change to `needAutoGenFields`

File: org.talend.mdm.core/src/com/amalto/core/load/context/DefaultStateContext.java [hunk 3]
```diff
@@ -62,25 +63,40 @@
 
     private String currentIdElementName;
 
-    public DefaultStateContext(String payLoadElementName, String[] idPaths, String dataClusterName, String dataModelName, int payloadLimit, LoadParserCallback callback) {
-        this(payLoadElementName, idPaths, dataClusterName, dataModelName, callback);
+    private AutoIdGenerator[] autoNormalFieldGenerator;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/context/DefaultStateContext.java]: Change to `normalFieldGenerator`

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java [hunk 4]
```diff
@@ -59,4 +75,59 @@
...
+
+    AutoIdGenerator[] getAutoFieldGenerator();
+
+    Stack<String> getReadElementPath();
+
+    /**
+     * judge the AUTO_INCREMENT/UUID field is existed in the supplied xml content.
+     * 1. if the field is in the entity,
+     *   1) supplied the value in the xml content: don't add
+     *   2) don't supplied in the xml content: added this field.
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: `not supplied`

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java [hunk 5]
```diff
@@ -59,4 +75,59 @@
...
+     * 2. the field is existed in one complex type such as Course/Like,
+     *   1). supplied in the xml content, don't add.
+     *   2). don't supplied in the xml content:
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: `not supplied`

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java [hunk 6]
```diff
@@ -59,4 +75,59 @@
...
+     *     I. contains other field's value of this complex type, added this field.
+     *     II. have no any fields'value of this complex type,  don't add.
+     * @return all fields of type are AUTO_INCREMENT or UUID field except PK, which need to generate value.
+     */
+    default String[] needAutoIncGeneratorField() {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: Change to `getAutoIncrementNormalFields`

File: org.talend.mdm.core/src/com/amalto/core/load/xml/AutoFieldGeneration.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        try {
+            AutoIdGenerator[] autoGenerator = context.getAutoFieldGenerator();
+            int i = 0;
+            /*
+            fielldPaths store the all need to generate value the field,  if doesn't existed, it means don't generate the value.
+            if the field path is 'Course/Score', it should be first write '<Course>', and next is '<Score>',
+            but it inverse for the end element, first is '<Score>', first is '<Course>'
+             */
+            for (String idPath : fieldPaths) {
+                if(idPath == null){
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/xml/AutoFieldGeneration.java]: Format the code

File: org.talend.mdm.core/src/com/amalto/core/load/xml/AutoFieldGeneration.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                    for (int j = idPathArray.length - 1; j >= 0; j--) {
+                        context.getWriter().writeEndElement(idPathArray[j]);
+                    }
+                } else {
+                    context.getWriter().writeEndElement(idPath);
+                }
+
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to generate automatic id");
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/xml/AutoFieldGeneration.java]: Lost original exception.

File: org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java [hunk 9]
```diff
@@ -308,10 +308,13 @@ public String getCauseMessage(Throwable throwable) {
...
         synchronized (itemsToUpdate) {
             if (!this.hasMetAutoIncrement) {
-                Collection<FieldMetadata> keyFields = document.getType().getKeyFields();
+                Collection<FieldMetadata> allFields = document.getType().getFields();
                 boolean hasAutoincrementKey = false;
-                for (FieldMetadata keyField : keyFields) {
-                    hasAutoincrementKey = EUUIDCustomType.AUTO_INCREMENT.getName().equals(keyField.getType().getName());
+                for (FieldMetadata keyField : allFields) {
+                    if (EUUIDCustomType.AUTO_INCREMENT.getName().equals(keyField.getType().getName())) {
+                        hasAutoincrementKey = true;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/save/SaverSession.java]: `hasAutoIncrementKey`

File: org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java [hunk 10]
```diff
@@ -106,9 +116,10 @@ protected void doPut(HttpServletRequest request, HttpServletResponse response) t
         String dataClusterName = request.getParameter(PARAMETER_CLUSTER);
         String typeName = request.getParameter(PARAMETER_CONCEPT);
         String dataModelName = request.getParameter(PARAMETER_DATAMODEL);
-        boolean needValidate = Boolean.valueOf(request.getParameter(PARAMETER_VALIDATE));
-        boolean needAutoGenPK = Boolean.valueOf(request.getParameter(PARAMETER_SMARTPK));
-        boolean insertOnly = Boolean.valueOf(request.getParameter(PARAMETER_INSERTONLY));
+        boolean needValidate = Boolean.parseBoolean(request.getParameter(PARAMETER_VALIDATE));
+        boolean needAutoGenPK = Boolean.parseBoolean(request.getParameter(PARAMETER_SMARTPK));
+        boolean needAutoGenAutoFields = request.getParameter(PARAMETER_SMARTFIELDS) == null ? true : false;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java]: `needAutoGenNormalFields`

File: org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java [hunk 11]
```diff
@@ -121,27 +132,50 @@ protected void doPut(HttpServletRequest request, HttpServletResponse response) t
...
+        writer.write("</body></html>"); //$NON-NLS-1$
+    }
+
+    protected void bulkLoadSave(String dataClusterName, String dataModelName, InputStream inputStream, LoadAction loadAction,
+            XSDKey keyMetadata, XSDKey autoFieldMetadata) throws ServletException {
+        XmlServer server;
+        try {
+            server = Util.getXmlServerCtrlLocal();
+        } catch (Exception e) {
+            throw new ServletException(e);
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java]: New exception would better to have self-defined message together with original exception object.

File: org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java [hunk 12]
```diff
@@ -263,6 +295,38 @@ private XSDKey getTypeKey(String dataModelName, String typeName) throws Exceptio
...
+                fieldTypesList.add(subField.getType().getName());
+            }
+        }
+        return new XSDKey(".", fieldsList.toArray(new String[] {}), fieldTypesList.toArray(new String[] {})); //$NON-NLS-1$
+    }
+
+    private void getFieldIntactName(FieldMetadata field, List<FieldMetadata> fieldList) {
+        if (field instanceof SimpleTypeFieldMetadata) {
+            String name = field.getType().getName();
+            if (EUUIDCustomType.AUTO_INCREMENT.getName().equals(name) || EUUIDCustomType.UUID.getName().equals(name)) { // See
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java]: Remove `// See`

File: org.talend.mdm.core/src/com/amalto/core/load/action/OptimizedLoadAction.java [hunk 13]
```diff
@@ -37,54 +37,91 @@
...
     public OptimizedLoadAction(String dataClusterName, String typeName, String dataModelName, boolean needAutoGenPK) {
+        this(dataClusterName, typeName, dataModelName, needAutoGenPK, false);
+    }
+
+    public OptimizedLoadAction(String dataClusterName, String typeName, String dataModelName, boolean needAutoGenPK, boolean needAutoGenNor){
         this.dataClusterName = dataClusterName;
         this.typeName = typeName;
         this.dataModelName = dataModelName;
         this.needAutoGenPK = needAutoGenPK;
+        this.needAutoGenFields = needAutoGenNor;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/action/OptimizedLoadAction.java]: `needAutoGenNor` to `needAutoGenFields`

File: org.talend.mdm.core/src/com/amalto/core/load/action/OptimizedLoadAction.java [hunk 14]
```diff
@@ -37,54 +37,91 @@
...
+                return new UUIDIdGenerator();
+            } else {
+                throw new UnsupportedOperationException(
+                        "No support for key field type '" + idFieldType + "' with autogen pk on."); //$NON-NLS-1$ //$NON-NLS-2$
+            }
+        }
+        return null;
+    }
+
+    private AutoIdGenerator[] getNormalFieldGenerator(XSDKey normalMetadata, boolean needAutoGen) {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/action/OptimizedLoadAction.java]: `getAutoFieldGenerator`

File: org.talend.mdm.core/src/com/amalto/core/load/context/AutoGenStateContext.java [hunk 15]
```diff
@@ -45,10 +49,26 @@
...
     private final AutoIdGenerator generator;
 
-    private AutoGenStateContext(StateContext delegate, String[] idPaths, AutoIdGenerator generator) {
+    private List<PathMatcher> normalFieldPaths;
+
+    private List<String> normalFieldInXML;
+
+    private Stack<String> readElementPath;
+
+    private final AutoIdGenerator[] autoNormalFieldGenerator;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/context/AutoGenStateContext.java]: to `normalFieldGenerator`

File: org.talend.mdm.core/src/com/amalto/core/load/context/DefaultStateContext.java [hunk 16]
```diff
@@ -62,25 +63,40 @@
 
     private String currentIdElementName;
 
-    public DefaultStateContext(String payLoadElementName, String[] idPaths, String dataClusterName, String dataModelName, int payloadLimit, LoadParserCallback callback) {
-        this(payLoadElementName, idPaths, dataClusterName, dataModelName, callback);
+    private AutoIdGenerator[] normalFieldGenerator;
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/context/DefaultStateContext.java]: to `normalFieldGenerators`

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java [hunk 17]
```diff
@@ -59,4 +75,59 @@
     boolean skipElement();
 
     void close(XmlServer server);
+
+    AutoIdGenerator[] getAutoFieldGenerator();
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: to `getNormalFieldGenerators`

File: org.talend.mdm.core/src/com/amalto/core/load/payload/FlushXMLReader.java [hunk 18]
```diff
@@ -116,6 +117,14 @@ public void parse(InputSource input) throws IOException, SAXException {
             while (!context.hasFinishedPayload()) {
                 context.parse(reader);
             }
+
+            String[] needAutoIncGeneratorField = context.getAutoIncrementNormalFields();
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/payload/FlushXMLReader.java]: to `autoIncrementNormalFields`

File: org.talend.mdm.core/src/com/amalto/core/load/xml/AutoFieldGeneration.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+    public AutoFieldGeneration(State previousState, String[] fieldPaths) {
+        this.previousState = previousState;
+        this.fieldPaths = fieldPaths;
+    }
+
+    @Override
+    public void parse(StateContext context, XMLStreamReader reader) throws XMLStreamException {
+        try {
+            AutoIdGenerator[] autoGenerator = context.getAutoFieldGenerator();
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/load/xml/AutoFieldGeneration.java]: `autoGenerator` to `normalFieldGenerators`

File: org.talend.mdm.core/src/com/amalto/core/save/context/BulkLoadSaver.java [hunk 20]
```diff
@@ -25,20 +25,24 @@
...
 
-    private final XSDKey keyMetadata;
+    private final XSDKey autoKeyMetadata;
+
+    private final XSDKey autoFieldMetadata;
 
     private final XmlServer server;
 
-    BulkLoadSaver(LoadAction loadAction, InputStream documentStream, XSDKey keyMetadata, XmlServer server) {
+    BulkLoadSaver(LoadAction loadAction, InputStream documentStream, XSDKey keyMetadata, XSDKey autoFieldMetadata,
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/save/context/BulkLoadSaver.java]: `keyMetadata` to `autoKeyMetadata`

File: org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java [hunk 21]
```diff
@@ -206,6 +234,11 @@ protected AtomicInteger getDbRequests(String dataClusterName) {
 
     protected LoadAction getLoadAction(String dataClusterName, String typeName, String dataModelName, boolean needValidate,
             boolean needAutoGenPK, boolean updateReport, String source) {
+        return getLoadAction(dataClusterName, typeName, dataModelName, needValidate, needAutoGenPK, updateReport, source, false);
+    }
+
+    protected LoadAction getLoadAction(String dataClusterName, String typeName, String dataModelName, boolean needValidate,
+            boolean needAutoGenPK, boolean updateReport, String source, boolean needAutoGenAutoFields) {
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java]: `needAutoGenAutoFields` to `needAutoGenNormalFields`

File: org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java [hunk 22]
```diff
@@ -263,6 +290,38 @@ private XSDKey getTypeKey(String dataModelName, String typeName) throws Exceptio
         return new XSDKey(".", fields, fieldTypes); //$NON-NLS-1$
     }
 
+    /**
+     * @param fieldList all field except the PK in one entity
```
Comment (lpw82gf) [inline org.talend.mdm.core/src/com/amalto/core/servlet/LoadServlet.java]: Method usage should be also added in the comment.

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java [hunk 23]
```diff
@@ -59,4 +75,59 @@
...
+                        PathMatch match = pathMatcher.match(s);
+                        if (match == PathMatch.FULL) {
+                            isMatched = true;
+                            isParentMatched = false;
+                            break;
+                        } else if (match == PathMatch.PARTIAL && i++ == pathArray.length - 2) {
+                            isParentMatched = true;
+                        }
+                    }
+                } else if (!pathMatchedStr.contains("/") && !path.contains("/")) {
```
Comment (hw-zhu) [inline org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: $NON-NLS-1$

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java [hunk 24]
```diff
@@ -59,4 +75,59 @@
...
+                    if (pathMatcher.match(path) == PathMatch.FULL) {
+                        isMatched = true;
+                        break;
+                    }
+                }
+            }
+            if ((!isMatched && !pathMatchedStr.contains("/")) || isParentMatched) {
+                normalFieldPathList.add(pathMatchedStr);
+            } else {
+                normalFieldPathList.add(null);
```
Comment (hw-zhu) [inline org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: do we must add NULL value here?

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java [hunk 25]
```diff
@@ -59,4 +75,59 @@
...
+            }
+        }
+        return normalFieldPathList.toArray(new String[] {});
```
Comment (hw-zhu) [inline org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: normalFieldPathList.toArray(new String[list.size()]);

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java (hunk 23 above)
Comment (zfkang-talend) [reply org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: Used the @SuppressWarnings("nls").

File: org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java (hunk 24 above)
Comment (zfkang-talend) [reply org.talend.mdm.core/src/com/amalto/core/load/context/StateContext.java]: yes, the index is correspondence with the array of the AutoIdGenerator.

File: org.talend.mdm.core/src/com/amalto/core/load/xml/StartElement.java [hunk 26]
```diff
@@ -30,6 +30,41 @@ private StartElement() {
...
+                Course/Like
+         */
+        StringBuilder name = new StringBuilder();
+        int i = 0;
+        for (String elementName : context.getReadElementPath()) {
+            if (i++ > 0) {
+                name.append("/").append(elementName);
+            }
+        }
+        name.delete(0, 1);
```
Comment (hw-zhu) [inline org.talend.mdm.core/src/com/amalto/core/load/xml/StartElement.java]: Try to the utils method Joiner.on("/") .join(context.getReadElementPath());

**Comment by zfkang-talend** [comment]
For SOAP, REST, and BulkLoad, they used the same Generator class to initial the AutoIdGenerator, but it needs to test in the cluster mode.

**Pull request merged**